src/policy/mallocspace/metadata.rs

```diff
@@ -1,61 +1,137 @@
+/// hence a separate `SideMetadata` instance is required.
+///
+/// This is a global side metadata spec even though it is used only by MallocSpace as
+/// we require its space to be contiguous and mapped only once. Otherwise we risk
+/// overwriting the previous mapping.
+pub(crate) const ACTIVE_CHUNK_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    is_global: true,
+    offset: GLOBAL_SIDE_METADATA_BASE_ADDRESS.as_usize(),
+    log_num_of_bits: 3,
+    log_min_obj_size: LOG_BYTES_IN_CHUNK as usize,
+};
+
 /// This is the metadata spec for the alloc-bit.
 ///
-/// An alloc-bit is required per min-object-size aligned address , rather than per object, and can only exist as side metadata.
+/// An alloc-bit is required per min-object-size aligned address, rather than per object, and can only exist as side metadata.
 ///
 /// The other metadata used by MallocSpace is mark-bit, which is per-object and can be kept in object header if the VM allows it.
 /// Thus, mark-bit is vm-dependant and is part of each VM's ObjectModel.
 ///
-#[cfg(target_pointer_width = "32")]
 pub(crate) const ALLOC_SIDE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
     is_global: false,
-    offset: 0,
+    offset: if cfg!(target_pointer_width = "64") {
+        LOCAL_SIDE_METADATA_BASE_ADDRESS.as_usize()
+    } else {
+        0
+    },
     log_num_of_bits: 0,
     log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
 };
 
+/// Metadata spec for the active page byte
+///
+/// We use a byte instead of a bit to avoid synchronization costs, i.e. to avoid
+/// the case where two threads try to update different bits in the same byte at
+/// the same time
 #[cfg(target_pointer_width = "64")]
-pub(crate) const ALLOC_SIDE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+pub(crate) const ACTIVE_PAGE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
     is_global: false,
-    offset: LOCAL_SIDE_METADATA_BASE_ADDRESS.as_usize(),
-    log_num_of_bits: 0,
-    log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
+    offset: ALLOC_SIDE_METADATA_SPEC.offset
+        + metadata_address_range_size(&ALLOC_SIDE_METADATA_SPEC),
+    log_num_of_bits: 3,
+    log_min_obj_size: constants::LOG_BYTES_IN_PAGE as usize,
+};
+
+#[cfg(target_pointer_width = "32")]
+pub(crate) const ACTIVE_PAGE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    is_global: false,
+    offset: ALLOC_SIDE_METADATA_SPEC.offset
+        + metadata_bytes_per_chunk(
+            ALLOC_SIDE_METADATA_SPEC.log_min_obj_size,
+            ALLOC_SIDE_METADATA_SPEC.log_num_of_bits,
+        ),
+    log_num_of_bits: 3,
+    log_min_obj_size: constants::LOG_BYTES_IN_PAGE as usize,
 };
 
 pub fn is_meta_space_mapped(address: Address) -> bool {
     let chunk_start = conversions::chunk_align_down(address);
-    ACTIVE_CHUNKS.read().unwrap().contains(&chunk_start)
+    is_chunk_mapped(chunk_start) && is_chunk_marked(chunk_start)
+}
+
+fn map_chunk_mark_space(chunk_start: Address) {
```

Comment:
It's clear to me what this function does, but I think a comment such as the ones inside the function body will be helpful here